core/libs/commonwealth/commonwealth/settings/tests/test_manager.py

```diff
@@ -0,0 +1,211 @@
+
+import pykson  # type: ignore
+
+from .. import manager, settings
+
+
+class SettingsV1(settings.BaseSettings):
+    VERSION = 1
+    version_1_variable = pykson.IntegerField(default_value=42)
+
+    def __init__(self, *args: str, **kwargs: int) -> None:
+        super().__init__(*args, **kwargs)
+
+        self.VERSION = SettingsV1.VERSION
+
+    def migrate(self, data: Dict[str, Any]) -> None:
+        if data["VERSION"] == SettingsV1.VERSION:
+            return
+
+        if data["VERSION"] < SettingsV1.VERSION:
+            super().migrate(data)
+
+        data["VERSION"] = SettingsV1.VERSION
+        data["version_1_variable"] = self.version_1_variable
+
+
+class SettingsV2(SettingsV1):
+    VERSION = 2
+    version_2_variable = pykson.IntegerField(default_value=66)
+
+    def __init__(self, *args: str, **kwargs: int) -> None:
+        super().__init__(*args, **kwargs)
+
+        self.VERSION = SettingsV2.VERSION
+
+    def migrate(self, data: Dict[str, Any]) -> None:
+        if data["VERSION"] == SettingsV2.VERSION:
+            return
+
+        if data["VERSION"] < SettingsV2.VERSION:
+            SettingsV1().migrate(data)
+
+        data["VERSION"] = SettingsV2.VERSION
+        data["version_2_variable"] = self.version_2_variable
+
+
+class SettingsV3(SettingsV2):
+    VERSION = 3
+    version_3_variable = pykson.IntegerField(default_value=99)
+
+    def __init__(self, *args: str, **kwargs: int) -> None:
+        super().__init__(*args, **kwargs)
+
+        self.VERSION = SettingsV3.VERSION
+
+    def migrate(self, data: Dict[str, Any]) -> None:
+        if data["VERSION"] == SettingsV3.VERSION:
+            return
+
+        if data["VERSION"] < SettingsV3.VERSION:
+            SettingsV2().migrate(data)
+
+        data["VERSION"] = SettingsV3.VERSION
+        data["version_3_variable"] = self.version_3_variable
+
+
+class SettingsV12(SettingsV3):
+    VERSION = 12
+    version_12_variable = pykson.IntegerField(default_value=1992)
+
+    def __init__(self, *args: str, **kwargs: int) -> None:
+        super().__init__(*args, **kwargs)
+
+        self.VERSION = SettingsV12.VERSION
+
+    def migrate(self, data: Dict[str, Any]) -> None:
+        if data["VERSION"] == SettingsV12.VERSION:
+            return
+
+        if data["VERSION"] < SettingsV12.VERSION:
+            SettingsV3().migrate(data)
+
+        data["VERSION"] = SettingsV12.VERSION
+        data["version_12_variable"] = self.version_12_variable
+
+
+def test_basic_settings_save_load() -> None:
+    temporary_folder = tempfile.mkdtemp()
+    config_path = pathlib.Path(temporary_folder)
+
+    # Test v1 save
+    settings_manager = manager.Manager("ManagerTest", SettingsV1, config_path)
+    settings_manager.settings.version_1_variable = 2022
+    settings_manager.save()
+
+    assert settings_manager.settings.version_1_variable == 2022
+
+    # Test v1 load
+    settings_manager = manager.Manager("ManagerTest", SettingsV1, config_path)
+    settings_manager.load()
```

Comment:
Would it make sense to put the `load()` call inside `__init__()`, making it unnecessary for the user to explicitly call it?

Reply:
Good call.